components/sync15/src/client_types.rs

```diff
@@ -24,60 +24,70 @@ pub struct RemoteClient {
     pub fxa_device_id: Option<String>,
     pub device_name: String,
     #[serde(default)]
-    #[serde(deserialize_with = "deserialize_option_device_type")]
-    #[serde(serialize_with = "serialize_option_device_type")]
-    pub device_type: Option<DeviceType>,
-}
-
-fn deserialize_option_device_type<'de, D>(d: D) -> std::result::Result<Option<DeviceType>, D::Error>
-where
-    D: serde::de::Deserializer<'de>,
-{
-    let opt_device_type: Option<DeviceType> = Option::deserialize(d)?;
-    Ok(match opt_device_type {
-        Some(val) => match val {
-            DeviceType::Unknown => None,
-            _ => Some(val),
-        },
-        None => None,
-    })
-}
-
-fn serialize_option_device_type<S>(id: &Option<DeviceType>, s: S) -> Result<S::Ok, S::Error>
-where
-    S: Serializer,
-{
-    match id {
-        Some(DeviceType::Unknown) => s.serialize_none(),
-        Some(d) => d.serialize(s),
-        None => s.serialize_none(),
-    }
+    pub device_type: DeviceType,
 }
 
 /// Enumeration for the different types of device.
 ///
 /// Firefox Accounts and the broader Sync universe separates devices into broad categories for
 /// display purposes, such as distinguishing a desktop PC from a mobile phone.
 
-#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
 pub enum DeviceType {
-    #[serde(rename = "desktop")]
     Desktop,
-    #[serde(rename = "mobile")]
     Mobile,
-    #[serde(rename = "tablet")]
     Tablet,
-    #[serde(rename = "vr")]
     VR,
-    #[serde(rename = "tv")]
     TV,
     // Unknown is a bit odd - it should never be set (ie, it's never serialized)
     // and exists really just so we can avoid using an Option<>.
-    #[serde(other)]
-    #[serde(skip_serializing)] // Don't you dare trying.
     Unknown,
 }
 
+impl Default for DeviceType {
+    fn default() -> Self {
+        DeviceType::Unknown
+    }
+}
+
+impl<'de> Deserialize<'de> for DeviceType {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        Ok(match String::deserialize(deserializer) {
+            Ok(s) => match s.as_str() {
+                "desktop" => DeviceType::Desktop,
+                "mobile" => DeviceType::Mobile,
+                "tablet" => DeviceType::Tablet,
+                "vr" => DeviceType::VR,
+                "tv" => DeviceType::TV,
+                _ => DeviceType::Unknown,
```

Comment:
Should strings that we don't know about map to `DeviceType::Unknown`?  It makes logical sense, but it seems like there could be issues with syncing:

  - Client A sets a field to `DeviceType::BrainImplant`
  - Client B is running an older FF version.  It syncs, doesn't recognize that variant, so it uses `DeviceType::Unknown`
  - Client B syncs again and uploads the record to the server
  - Client A syncs, updates their value to `DeviceType::Unknown`

I'm not sure this can happen right now, but maybe something like that could happen in the future.  Maybe it would be better to throw an error in that case and make client B ignore the record.  I'm sure what's best, so whatever you think is fine with me.  It should be documented in any case though.

Reply:
I'm slightly partial to how it is since that device type technically is "unknown" at that point, and any other way might get a little messy trying to support old formats, etc. However, I do agree documenting it would help us down the line.